File: packages/coreutils/src/commands/rm.rs
use super::{write_stderr, remove_file, remove_dir, file_kind, read_dir, FileKind};

pub fn run(args: &[&str]) -> u8 {
    let mut recursive = false;
    let mut force = false;
    let mut file_args: Vec<&str> = Vec::new();

    for &arg in args {
        match arg {
            "-r" | "-R" | "--recursive" => recursive = true,
            "-f" | "--force" => force = true,
            "-rf" | "-fr" => { recursive = true; force = true; }
            a if a.starts_with('-') => {
                for c in a[1..].chars() {
                    match c {
                        'r' | 'R' => recursive = true,
                        'f' => force = true,
                        _ => {}
                    }
                }
            }
            _ => file_args.push(arg),
        }
    }

    let mut errors = 0u8;
    for &arg in &file_args {
        match file_kind(arg) {
            FileKind::Regular | FileKind::Other => {
                remove_file(arg);
            }
            FileKind::Directory => {
                if recursive {
                    remove_dir_recursive(arg);
                } else {
                    write_stderr(&format!("rm: cannot remove '{}': Is a directory\n", arg));
                    errors = 1;
                }
            }
            FileKind::NotFound => {
                if !force {
                    write_stderr(&format!("rm: cannot remove '{}': No such file or directory\n", arg));
                    errors = 1;
                }
            }
        }
    }
    errors
}

fn remove_dir_recursive(path: &str) {
    let entries = read_dir(path);
    for entry in entries {
        let child = format!("{}/{}", path.trim_end_matches('/'), entry);
        match file_kind(&child) {
            FileKind::Directory => remove_dir_recursive(&child),
            _ => { remove_file(&child); }
        }
    }
    remove_dir(path);
}

File: packages/coreutils/src/commands/rm.ts
/**
 * `rm` — remove files and directories.
 *   -r / -R / --recursive : remove directories and their contents recursively
 *   -f / --force          : ignore nonexistent operands, never error on missing
 *   -d / --dir            : remove empty directories
 *   -v / --verbose        : print each removed path
 */
import { defineCommand, parseArgs, writeLine } from '../harness.ts';
import { unlink, rmdir, readdir, typeOf, joinPath, normalize } from '../fs.ts';
import type { CommandFn, CommandIO } from '../harness.ts';

const rmCommand: CommandFn = async (io: CommandIO): Promise<number> => {
  const { positionals, flags } = parseArgs(io.args.slice(1), {
    boolean: ['r', 'R', 'f', 'd', 'v'],
    alias: { recursive: 'r', force: 'f', dir: 'd', verbose: 'v' },
  });
  const recursive = Boolean(flags.r) || Boolean(flags.R);
  const force = Boolean(flags.f);
  const out = io.stdout.getWriter();
  const err = io.stderr.getWriter();
  let code = 0;

  try {
    if (positionals.length === 0) {
      if (force) return 0;
      await writeLine(err, 'rm: missing operand');
      return 1;
    }
    for (const target of positionals) {
      const path = normalize(target);
      try {
        const t = await typeOf(io, path, false);
        if (t === undefined) {
          if (!force) { await writeLine(err, `rm: cannot remove '${target}': No such file or directory`); code = 1; }
          continue;
        }
        if (t === 'directory') {
          if (recursive) {
            await removeTree(io, path, Boolean(flags.v), out);
          } else if (flags.d) {
            await rmdir(io, path);
            if (flags.v) await writeLine(out, `removed directory '${target}'`);
          } else {
            await writeLine(err, `rm: cannot remove '${target}': Is a directory`);
            code = 1;
          }
        } else {
          await unlink(io, path);
          if (flags.v) await writeLine(out, `removed '${target}'`);
        }
      } catch (e) {
        if (!force) { await writeLine(err, `rm: cannot remove '${target}': ${(e as Error).message}`); code = 1; }
      }
    }
    return code;
  } finally {
    await out.close().catch(() => {});
    await err.close().catch(() => {});
  }
};

/** Depth-first remove a directory and everything beneath it. */
async function removeTree(
  io: CommandIO, path: string, verbose: boolean, out: WritableStreamDefaultWriter<Uint8Array>,
): Promise<void> {
  const entries = await readdir(io, path);
  for (const entry of entries) {
    const child = joinPath(path, entry.name);
    if (entry.type === 'directory') {
      await removeTree(io, child, verbose, out);
    } else {
      await unlink(io, child);
      if (verbose) await writeLine(out, `removed '${child}'`);
    }
  }
  await rmdir(io, path);
  if (verbose) await writeLine(out, `removed directory '${path}'`);
}

export default defineCommand(rmCommand);
export { rmCommand };
